### DISCUSSION TITLE
Mount NAS shared folder root as top level directory

### DISCUSSION THREAD

**Discussion by HylianPeasant** [retronas/discussions#23] [Q&A] [answered]
Hello,

I'm trying to achieve the following with retronas : leave retronas handling all the service and protocol stuff and keep the storage part on my NAS (where my roms and isos are located basically).

I'm successfully mounting retronas shared folder on the MiSTer FPGA and I'm also successfully mounting and accessing files on my NAS through retronas on /mnt/ for the moment.

What I would like to perform basically is defining my NAS shared folder as the top level directory.
But when I try to do that, folders on my NAS (games and saves) replace all folders on the top level directory.
Plus, my MiSTer is unable to mount nothing when my NAS shared folder is mounted as the top level directory.

On the technical part, a shared folder on my NAS is CIFS mounted on retronas and retronas itself is acting as a CIFS server and Mister CIFS mount the mister shared folder.

Obviously, the part where I'm trying to CIFS mount inside another CIFS mount doesn't help and is probably the obvious culprit.

Please, any idea on how to proceed to achieve this goal with a clean and proper form ?

HP

**Comment by sairuk** [comment, marked as answer]
The CIFS mount into RetroNAS will not be an issue in itself, this is an acceptable situation.

When you mount in a drive at a point the inode is remapped to the mounted device so your old files no longer showing is expected behaviour by design.

For this situation to work (providing i've read it properly) you will need
- mount in your NAS cifs share and point RetroNAS Top Level Directory to the mount point
- ensure the mounted data is rw for the RetroNAS user or Retronas is configured to use a user with rw permissions
- let RetroNAS build its required folder structure on the mounted drive (now Top Level Directory)

after this it should work relatively transparently

**Reply by HylianPeasant** [reply]
Hello sairuk,

Thanks for the reply.

EDIT 1: I'm still checking things on the client side to understand why I can't create symlinks on my CIFS mount.

EDIT 2: OK I had to directly mounting my CIFS share by editing the fstab. also had to add a new option : "mfsymlinks". Without this argument (mfsymlinks), symlinking doesn't work. Autofs doesn't seem to properly handle this option.
Need to do more checks to see if everything work fine.
If you can use NFS mount to your NAS, this is recommended as CIFS seems to have a few drawbacks when talking about symlinks.

Did what you instructed : 
1/ backup everything from my NAS shared folder in prep for below
2/ deleted everything already existing from /data/retronas/
3/ mounted my NAS shared folder on /data/retronas/ (which is the current top level directory)
4/ installed again Samba and MiSTer services

Here are the results :
1/ every template folder have been created, I can browse them locally on Retronas and they also show on my NAS shared folder (good)
2/ when I mount Retronas shared folder on MiSTer, this time one folder show up : the "games" folder (good)
3/ no symlinks are created : when I do a "find . type l -ls" command on /data/retronas/ no symlinks are reported, During the playbook excuting, a lot of lines are red (bad)

I've enabled "allow symbolic links within shared folders" option on the NAS samba service side with the same result.
I should add, I'm using autofs on Retronas to configure the cifs mounting part on the NAS side.

**Reply by sairuk** [reply]
yeh symlinking isn't a 'MS thing' well not until recently i believe, its works in some capacity. I'm not a windows/cifs user so yeh kinda hard to confirm stuff.

We did find the mfsymlinks thing in a ticket somewhere now that you've reminded me.

Ahhhh ... here it is for reference https://github.com/danmons/retronas/issues/106#issuecomment-1046179193

**Reply by HylianPeasant** [reply]
I was using NFS in the past but an update on my NAS messed with how I used group id to allow access to shared folder for users.
And since my NAS is a on the shelf solution (Synology), I didn't want to mess directly with uid/gid inside the server in order to make users and groups match across my systems and potentially break everything on my NAS.

It was easier to transition to Samba. While slightly poor in performance, it was quicker to manage plus the added security of password and data encryption that we get with NFS only when he install a Kerberos server for auth and encryption.

Maybe someday if I find time, I can try to make Kerberos work to get rid of CIFS shares.

**Reply by sairuk** [reply]
one of those things i guess, QNAP screwed NFS and i transitioned to cifs for a while then when they fixed it transitioned back. Terramaster have a cruddy UI implementation of NFS and you need to get into the console to actually do anything of use with it.

**Comment by HylianPeasant** [comment]
I will eventually configure nfs+kbr5p one day and get rid of cifs for my linux systems.

I can confirm that everything work like a charm now with my NAS shared folder mounted inside Retronas and Retronas itself acting as a CIFS server for my MiSTer with games, saves and savestates folders mounted correctly (might be a good idea to enable those two last folders by default or leave a more visible option to give the choice).

Now is time to toy a little bit with rom import :)

**Reply by drinksomewhiskey** [reply]
Hi, would you please be able to explain exactly how you made this work? I have tried but failed due to permissions issues. 

Essentially, I enabled NFS share on a shared folder on the Synology NAS allowing read/write and giving connected users root access. I was able to mount this via cockpit and point the directory to the NFS share in retronas. However, anytime I try to install any service via retronas I get failures do to permissions. 

Are there any steps I am missing? Thank you in advance.

**Reply by HylianPeasant** [reply]
Nothing really special. I'm also using Debian but with nfsv4.
Uid/gid match on both side (nfsv4 still rely on that).

#installation the nfs package
apt install nfs-common

#enable nfsv4: edit the file /etc/default/nfs-common and modify the following line
NEED_IDMAPD=yes

#enable nfsv4, edit the file /etc/idmap.conf and modify the following line to match the domain configured on the synology nfs server side
Domain = DOMAIN_NAME

#mount of the nfsv4 share
mount -t nfs4 server-ip-address:/share-path /mount-path

#mount of the nfsv4 share during the system startup, edit the file /etc/fstab and add the following line
server-ip-address:/share-path	/mount-path	nfs	nfsvers=4,sec=sys	0	0

**Reply by drinksomewhiskey** [reply]
> Nothing really special. I'm also using Debian but with nfsv4.
> 
> Uid/gid match on both side (nfsv4 still rely on that).
> 
> 
> 
> #installation the nfs package
> 
> apt install nfs-common
> 
> 
> 
> #enable nfsv4: edit the file /etc/default/nfs-common and modify the following line
> 
> NEED_IDMAPD=yes
> 
> 
> 
> #enable nfsv4, edit the file /etc/idmap.conf and modify the following line to match the domain configured on the synology nfs server side
> 
> Domain = DOMAIN_NAME
> 
> 
> 
> #mount of the nfsv4 share
> 
> mount -t nfs4 server-ip-address:/share-path /mount-path
> 
> 
> 
> #mount of the nfsv4 share during the system startup, edit the file /etc/fstab and add the following line
> 
> server-ip-address:/share-path	/mount-path	nfs	nfsvers=4,sec=sys	0	0

This was very helpful. My issue was related to the uid and gid mismatch as well as the user mapping on the Synology side, which I turned off. Thank you for all of your help.
